File: LAB_9/Proxy.cpp
#include <iostream>
#include <string>

// Interfejs dostępu do danych
class DataAccess {
    virtual void accessData(std::string user) = 0;
};

// Klasa rzeczywista (RealSubject)
class RealDataAccess : public DataAccess {
public:
    void accessData(std::string user) override {
        std::cout << "Access granted for user: " << user << std::endl;
    }
};

// Klasa Proxy
class DataAccessProxy : public DataAccess {
private:
    RealDataAccess* realDataAccess;
    const std::string AUTHORIZED_USER = "admin";

public:
    DataAccessProxy() {
        realDataAccess = new RealDataAccess();
    }

    void accessData(std::string user) override {
        std::cout << "Accessing data by: " << user << "..." << std::endl;

        if (isAuthorized(user)) {
            realDataAccess->accessData(user);
        } else {
            std::cout << "Access denied for user: " << user << std::endl;
        }
    }

    bool isAuthorized(std::string user) {
        return user == AUTHORIZED_USER;
    }
};

// Klient
int main() {
    DataAccessProxy* dataAccessProxy = new DataAccessProxy();
   
    // Dostęp dozwolony
    dataAccessProxy->accessData("admin");

    std::cout << std::endl;
    
    // Dostęp zabroniony
    dataAccessProxy->accessData("user");

    return 0;
}

File: LAB_9/Proxy.java
// Interfejs dostępu do danych
interface DataAccess {
    void accessData(String user);
}

// Klasa rzeczywista (RealSubject)
class RealDataAccess implements DataAccess {
    @Override
    public void accessData(String user) {
        System.out.println("Access granted for user: " + user);
    }
}

// Klasa Proxy
class DataAccessProxy implements DataAccess {
    private final RealDataAccess realDataAccess;
    private static final String AUTHORIZED_USER = "admin";

    public DataAccessProxy() {
        this.realDataAccess = new RealDataAccess();
    }

    @Override
    public void accessData(String user) {
        System.out.println("Accessing data by: " + user + "...");

        if (isAuthorized(user)) {
            realDataAccess.accessData(user);
        } else {
            System.out.println("Access denied for user: " + user);
        }
    }

    private boolean isAuthorized(String user) {
        return AUTHORIZED_USER.equals(user);
    }
}

// Klient
public class Proxy {
    public static void main(String[] args) {
        DataAccess dataAccess = new DataAccessProxy();

        // Dostęp dozwolony
        dataAccess.accessData("admin");

        System.out.println();

        // Dostęp zabroniony
        dataAccess.accessData("user");
    }
}